src/coreclr/jit/gentree.cpp

```diff
@@ -15411,6 +15459,20 @@ bool GenTree::DefinesLocal(Compiler* comp, GenTreeLclVarCommon** pLclVarTree, bo
             blkNode = AsOp()->gtOp1->AsBlk();
         }
     }
+    else if (OperIs(GT_CALL))
+    {
+        GenTreeLclVar* retBufLclVar = AsCall()->GetRetBufArg();
+        if (retBufLclVar != nullptr)
+        {
+            *pLclVarTree = retBufLclVar;
+
+            if (pIsEntire != nullptr)
+            {
+                *pIsEntire = false;
```

Comment:
It is still not clear to me why are we making the definition here partial.

They should be partial if we have a field (at a non-zero offset), but they should be full if we don't. It is a pessimization to make all of them partial.

Reply:
Hhm, my understanding was that if both `DEF` and `USEASG` is set, it is considered as partial def. (e.g. in `RenameDef()` ) 

https://github.com/dotnet/runtime/blob/ac246b17f3ecb0a2b32a74cf3a74b6c749065304/src/coreclr/jit/ssabuilder.cpp#L769-L783

But I guess since `fgAssignSetVarDef` will be used to set these flags, it will be taken care. I will make it "full definition".

Reply:
Note that the logic for "partial-vs-full" must consider what struct the call originally returned, and the kind of local address we're storing to.

I. e. for this method it would ideally look like this:
```c
if (OperIs(GT_CALL))
{
    GenTree* addr = call->GetLclRetBufArgNode(); // Note it must return the whole "ADDR(LCL)" tree.
    unsigned size = typGetObjLayout(call->gtRetClsHnd);
    return addr->DefinesLocalAddr(comp, size, pLclVarTree, pIsEntire);
}
```
Since this `GetLclRetBufArgNode` method is used above, I'll also outline what I came to with https://github.com/dotnet/runtime/pull/64130#discussion_r814962297, after considering a few options (I am not incredibly happy with the end result, but the other options were worse in one way or another):
1) In local address visitor, save the `Use*` for the return buffer argument. The presence of this `Use*` now marks the call as potentially defining a tracked local.
2) In morph, update the `Use*` if the return buffer is moved to the late arg list with the late use in case it was moved "fully" (i. e. replaced by an `ARGPLACE`), leave as-is if it was moved via a setup arg.
3) After morph, use `GetLclRetBufArgNode`, that will roughly look like this:
```c
// I really don't like how we create a dependency on the setup args here, but hope
// to eliminate it in the future by using local address nodes instead of ADDRs here, always.
assert(use != nullptr);
return use->GetNode->OperIs(GT_ASG) ? use->GetNode()->gtGetOp2() : use->GetNode();
```
To get the address node and use it in `DefinesLocal`.

As I mentioned, this is not "the best it could be" (the best would be to represent these cases with an actual `ASG`), but "better" in that it aligns as much as reasonable with how we treat ordinary stores.

Reply:
This markes sense.